refactor: Refactor Meshing workflow. (#2433)
ansys/pyfluent: doc/source/user_guide/new_meshing_workflows.rst

.. _ref_user_guide_new_meshing_workflows:

> Use meshing workflows
> =====================

Reviewer: ```suggestion
Using meshing workflows
=======================
```
Please apply the same change in the other file.

> PyFluent supports accessing all Fluent meshing functionalities, including
> guided meshing workflows.

Reviewer: ```suggestion
Fluent meshing guided workflows are designed to make it easier to generate a volume mesh starting from CAD geometries. You can use PyFluent to access the guided workflows.
```
Please: other file too.

Watertight geometry meshing workflow
------------------------------------

> This simple example shows how to use the watertight geometry meshing workflow.

Reviewer: ```suggestion
Use the **Watertight Geometry** workflow for water-tight CAD geometries that do not require much in the way of clean-up or modifications. The following example shows you how to use the Watertight Geometry workflow.
```

Import geometry
~~~~~~~~~~~~~~~

.. code:: python

    import ansys.fluent.core as pyfluent
    from ansys.fluent.core import examples

    import_file_name = examples.download_file('mixing_elbow.pmdb', 'pyfluent/mixing_elbow')
    meshing = pyfluent.launch_fluent(
        mode="meshing", precision='double', processor_count=2
    )
    watertight = meshing.watertight()
    watertight.import_geometry.FileName.set_state(import_file_name)
    watertight.import_geometry.LengthUnit.set_state('in')

>     watertight.import_geometry.Execute()

Reviewer: ```suggestion
    watertight.import_geometry()
```
This should work and it is the intended usage. We could eliminate `Execute` in this mode.

Add local sizing
~~~~~~~~~~~~~~~~

.. code:: python

>     watertight.add_local_sizing.AddChildToTask()

Reviewer: We don't want to expose anything like this, but there is something to be aware of, as follows. We have code like the following in the meshing workflow tests. But this relies on AddChild which is a specific argument in meshing datamodel.
```
    add_local_sizing = watertight.add_local_sizing
    add_local_sizing.add_child(state={"BOIFaceLabelList": ["cold-inlet"]})
```
This is the code in workflow.py. As I say AddChild is specific to meshing.py:
```
    def add_child(self, state: Optional[dict] = None) -> None:
        """Add a child to this CompoundTask.

        Parameters
        ----------
        state : Optional[dict]
            Optional state.
        """
        state = state or {}
        state.update({"AddChild": "yes"})
        self._task.Arguments.set_state(state) 
```
and then
```
    def add_child_and_update(self, state=None):
        """Add a child to this CompoundTask and update.

        Parameters
        ----------
        state : Optional[dict]
            Optional state.
        """
        self.add_child(state)
        self._task.AddChildAndUpdate()
        return self.last_child()
```
Ignoring for a moment the coupling with the meshing data model, you should look into hiding `AddChildToTask` just as for `AddChildAndUpdate`. `AddChildToTask` is simpler.

Describe geometry
~~~~~~~~~~~~~~~~~

.. code:: python

    watertight.describe_geometry.UpdateChildTasks(SetupTypeChanged=False)
    watertight.describe_geometry.SetupType.set_state("The geometry consists of only fluid regions with no voids")
    watertight.describe_geometry.UpdateChildTasks(SetupTypeChanged=True)
    watertight.describe_geometry.Execute()

Update boundaries
~~~~~~~~~~~~~~~~~

.. code:: python

    watertight.update_boundaries.BoundaryLabelList.set_state(["wall-inlet"])
    watertight.update_boundaries.BoundaryLabelTypeList.set_state(["wall"])
    watertight.update_boundaries.OldBoundaryLabelList.set_state(["wall-inlet"])
    watertight.update_boundaries.OldBoundaryLabelTypeList.set_state(["velocity-inlet"])
     watertight.update_boundaries.Execute()

Update regions
~~~~~~~~~~~~~~

.. code:: python

    watertight.update_regions.Execute()

Add boundary layers
~~~~~~~~~~~~~~~~~~~

.. code:: python

    watertight.add_boundary_layer.AddChildToTask()
    watertight.add_boundary_layer.InsertCompoundChildTask()
    meshing.workflow.TaskObject["smooth-transition_1"].Arguments = {
        "BLControlName": "smooth-transition_1",
    }

>     watertight.add_boundary_layer.Arguments = {}

Reviewer: `Arguments` should not be exposed in this mode. it should be `arguments` for getting and for setting.

    watertight.task("smooth-transition_1")()

Generate volume mesh
~~~~~~~~~~~~~~~~~~~~

.. code:: python

    watertight.create_volume_mesh.VolumeFill.set_state("poly-hexcore")
    watertight.create_volume_mesh.VolumeFillControls.HexMaxCellLength.set_state(0.3)
    watertight.create_volume_mesh()

Switch to solution mode
~~~~~~~~~~~~~~~~~~~~~~~

.. code:: python

    solver = meshing.switch_to_solver()

Fault-tolerant meshing workflow
-------------------------------

> This simple example shows how to use the fault-tolerant meshing workflow.

Reviewer: ```suggestion
Use the **Fault-tolerant** meshing workflow for more complicated non-water-tight CAD geometries that may require some form of clean-up or modification (for example, defects such as overlaps, intersections, holes, duplicates, etc). This simple example shows how to use the fault-tolerant meshing workflow.
```
